### ISSUE TITLE
MDS: slow and broken pipe crash

### ISSUE THREAD

**Issue by ajdapretnar** [biolab/orange3#5356]
- [ ] **What's wrong?**

In the recent release (3.28.0), MDS is incredibly slow when randomizing data and trying to run the optimization on iris.
But a worse problem is that it crashes with Broken Pipe:

```
Exception: | BrokenPipeError: [Errno 32] Broken pipe
-- | --
Module: | None
Version: | 3.28.0
Environment: | Python 3.8.8 on Darwin 17.7.0 Darwin Kernel Version 17.7.0: Fri Oct 30 13:34:27 PDT 2020; root:xnu-4570.71.82.8~1/RELEASE_X86_64 x86_64
Installed Packages: | AnyQt==0.0.11, Bottleneck==1.3.2, CacheControl==0.12.6, Orange3-Educational==0.4.0, Orange3-Explain==0.2.1, Orange3-Geo==0.2.8, Orange3-ImageAnalytics==0.6.0, Orange3-Text==1.3.1, Orange3==3.28.0, Pillow==8.1.2, PyQt5-sip==12.8.1, PyQt5==5.12.2, PyQtWebEngine==5.12.1, PyYAML==5.4.1, Shapely==1.7.1, XlsxWriter==1.3.7, appnope==0.1.2, baycomp==1.0.2, beautifulsoup4==4.9.3, biopython==1.78, catboost==0.24.4, certifi==2020.12.5, chardet==3.0.4, click==7.1.2, commonmark==0.9.1, cycler==0.10.0, decorator==4.4.2, defusedxml==0.7.1, dictdiffer==0.8.1, docutils==0.16, docx2txt==0.8, et-xmlfile==1.0.1, gensim==3.8.3, graphviz==0.16, h11==0.12.0, httpcore==0.12.3, httpx==0.16.1, idna==2.10, joblib==0.11, keyring==10.3.1, keyrings.alt==2.2, kiwisolver==1.3.1, llvmlite==0.36.0, lockfile==0.12.2, lxml==4.6.2, matplotlib==3.3.4, msgpack==1.0.2, ndf==0.1.4, networkx==2.5, nltk==3.5, numba==0.53.0, numpy==1.19.5, oauthlib==3.1.0, odfpy==1.4.1, openTSNE==0.4.4, openpyxl==3.0.7, orange-canvas-core==0.1.19, orange-widget-base==4.12.0, pandas==1.1.5, pdfminer3k==1.3.4, pip==19.3.1, plotly==4.14.3, ply==3.11, pyparsing==2.4.7, pyqtgraph==0.11.0, python-dateutil==2.8.1, python-louvain==0.15, pytz==2021.1, qasync==0.13.0, regex==2021.3.17, requests-oauthlib==1.3.0, requests==2.25.1, retrying==1.3.3, rfc3986==1.4.0, scikit-learn==0.23.2, scipy==1.5.4, serverfiles==0.3.1, setuptools-git==1.2, setuptools==49.2.1, shap==0.37.0, simhash==2.0.0, simplejson==3.17.2, six==1.15.0, slicer==0.0.3, smart-open==4.2.0, sniffio==1.2.0, soupsieve==2.2.1, threadpoolctl==2.1.0, tqdm==4.59.0, tweepy==3.10.0, ufal.udpipe==1.2.0.3, urllib3==1.26.3, wheel==0.36.2, wikipedia==1.4.0, xgboost==1.3.3, xlrd==1.2.0
Machine ID: | ajda
Stack Trace: | BrokenPipeError: [Errno 32] Broken pipe
```

- [ ] **How can we reproduce the problem?** 

File (iris) - MDS. Randomize points and try to run the plot with update on every iteration.

- [ ] **What's your environment?**

- Operating system: OSX
- Orange version: 3.28.0
- How you installed Orange: dmg

**Comment by markotoplak** [comment]
Maybe connected, we keep getting random BrokenPipeErrors outside MDS on a workshop. I have not been as diligent as Ajda to have a reproducible version; they seemed so random for now...

**Comment by markotoplak** [comment]
Just a note, before a forget. These BrokenPipes were happening on a Mac. On one user's Wins computer I saw equally unexplainable and strange OSError. Maybe something separate, but maybe these things are connected and show as different errors.

**Comment by irgolic** [comment]
@markotoplak, you mentioned a lot of things were upgraded in this release. Could you list them?

**Comment by markotoplak** [comment]
They are in the CHANGELOG.md. But I do not know what could cause this. At first, I attributed this to the new restart loop (because Feri used clear widget settings which restarted his Orange), then I thought about multithreaded warnings handler, but I do not really remember else anything changing the underlying architecture so strongly.

Also, there were some changes to widget-base and canvas-core. :)

**Comment by irgolic** [comment]
Perhaps I remember incorrectly, in addition to the many PRs, did you not mention that we're distributing a newer python or PyQt?

**Comment by markotoplak** [comment]
@irgolic, yes, also, a newer PyQtGraph had some performance regression with scatterplots (I never measured, just something they talked about on github).

**Comment by j9ac9k** [comment]
I felt a disturbance in the force.  Our (pyqtgraph) performance regression had to do w/ enabling a transparent background, which affected scatter plot performance pretty heavily (I understand Orange3 does this explicitly, so there should have been no noticeable performance hit here).  Our scatter plot code has actually seen significant performance improvements since 0.11.0.

Regarding this issue, the _closest_ thing that I can think of is we sometimes get broken pipe errors in our remote plotting examples, but I Think they have to do with how the examples are closed/terminated.

I should also point out we stopped testing pyqtgraph against conda sources Qt bindings, not sure if you all have something to replicate, but if you can, my first suggestion would be to swap out the conda packaged Qt bindings with something provided by pip/pypi.

**Comment by markotoplak** [comment]
@j9ac9k, thank you for the comment! These problems actually appeared on pip-installed PyQt on Mac... I didn't go deep enough into it to have an idea what is causing it, we are just listing possibilities.

Now I checked: the problematic package still has `pyqtgraph==0.11.0`, which predates pyqtgraphs's scatterplot optimizations. Sorry for the disturbance. :)

**Comment by j9ac9k** [comment]
No problem; feel free to ping me, although I make no promises regarding how helpful I would be 😆

**Comment by ajdapretnar** [comment]
Just checked with PyQt5==5.14.2, the issue still remains. I can describe it better, not even sure it is related to PyQt.
- Press Randomize in MDS. It takes forever (even for iris).
- Set Refresh to "Every iteration".
- Run optimization. The visualization does not refresh at every iteration, it just shows the final result.

p.s. I have not gotten the BrokenPipe crash on master, just on dmg version.
